File: frontend/src/lib/phases.ts
/**
 * Canonical phase definitions.
 *
 * Every place in the frontend that needs the phase order, the short
 * label, or a route to a phase page imports from here. Never duplicate
 * these lists.
 */

export const PHASE_ORDER = [
  "idea",
  "plan",
  "blueprint",
  "scaffold",
  "development",
  "deployment",
  "maintenance",
] as const;

export type PhaseId = (typeof PHASE_ORDER)[number];

/**
 * Short label used in nav, chips, and jump links.
 * Keep these to one or two words.
 */
export const PHASE_LABEL: Record<PhaseId, string> = {
  idea: "Idea",
  plan: "Plan",
  blueprint: "Blueprint",
  scaffold: "Scaffold",
  development: "Dev",
  deployment: "Deploy",
  maintenance: "Maintenance",
};

/**
 * Long label used in page headers and body copy.
 */
export const PHASE_LABEL_LONG: Record<PhaseId, string> = {
  idea: "Idea",
  plan: "Plan",
  blueprint: "Blueprint",
  scaffold: "Scaffold",
  development: "Development",
  deployment: "Deployment",
  maintenance: "Maintenance",
};

/**
 * Ordinal position of each phase, starting at 1. Used in page headers
 * ("Phase 3 — Blueprint").
 */
export const PHASE_NUMBER: Record<PhaseId, number> = {
  idea: 1,
  plan: 2,
  blueprint: 3,
  scaffold: 4,
  development: 5,
  deployment: 6,
  maintenance: 7,
};

export function isPhase(value: string): value is PhaseId {
  return (PHASE_ORDER as readonly string[]).includes(value);
}

export function phaseLabel(phase: string): string {
  return isPhase(phase) ? PHASE_LABEL[phase] : phase;
}

export function phaseLabelLong(phase: string): string {
  return isPhase(phase) ? PHASE_LABEL_LONG[phase] : phase;
}

export function phaseNumber(phase: string): number | null {
  return isPhase(phase) ? PHASE_NUMBER[phase] : null;
}

export function phaseUrl(projectSlug: string, phase: string): string {
  return `/projects/${projectSlug}/${phase}`;
}

File: backend/app/core/phases.py
"""
Canonical phase definitions for SOJIP.

Every place in the backend that needs to know the phase order, the
human-readable label, or which artifact kind belongs to a phase
imports from this module. Never duplicate these lists.
"""
from typing import Literal

PhaseId = Literal[
    "idea",
    "plan",
    "blueprint",
    "scaffold",
    "development",
    "deployment",
    "maintenance",
]


# Canonical order. Anything that walks the journey uses this.
PHASE_ORDER: tuple[str, ...] = (
    "idea",
    "plan",
    "blueprint",
    "scaffold",
    "development",
    "deployment",
    "maintenance",
)


# Human-readable labels — used in prompts, error messages, and
# anywhere the backend needs to talk about a phase by name.
PHASE_LABEL: dict[str, str] = {
    "idea": "Idea",
    "plan": "Plan",
    "blueprint": "Blueprint",
    "scaffold": "Scaffold",
    "development": "Development",
    "deployment": "Deployment",
    "maintenance": "Maintenance",
}


# Which artifact kind belongs to each phase, if any.
# `None` means the phase produces no artifact (the work IS the artifact).
PHASE_ARTIFACT_KIND: dict[str, str | None] = {
    "idea": "lean_canvas",
    "plan": "task_graph",
    "blueprint": "system_diagram",
    "scaffold": None,
    "development": None,
    "deployment": None,
    "maintenance": "retrospective",
}


def is_valid_phase(phase: str) -> bool:
    return phase in PHASE_ORDER


def next_phase(current: str) -> str | None:
    """Return the phase after `current`, or None if at the end."""
    try:
        idx = PHASE_ORDER.index(current)
    except ValueError:
        return None
    if idx + 1 >= len(PHASE_ORDER):
        return None
    return PHASE_ORDER[idx + 1]


def previous_phase(current: str) -> str | None:
    """Return the phase before `current`, or None if at the start."""
    try:
        idx = PHASE_ORDER.index(current)
    except ValueError:
        return None
    if idx == 0:
        return None
    return PHASE_ORDER[idx - 1]


def artifact_kind_for(phase: str) -> str | None:
    return PHASE_ARTIFACT_KIND.get(phase)


def label_for(phase: str) -> str:
    return PHASE_LABEL.get(phase, phase)
